src/client/libds3/bucket.c

```diff
@@ -7,31 +7,409 @@
+	daos_size_t                 bi    = 0;
+	ds3_bucket_t               *ds3b  = NULL;
+
+	if (ds3 == NULL || nbuck == NULL || buf == NULL || marker == NULL)
+		return -EINVAL;
+
+	D_ALLOC_ARRAY(conts, ncont);
+	if (conts == NULL) {
+		return -ENOMEM;
+	}
+
+	// TODO: Handle markers
+	rc = daos_pool_list_cont(ds3->poh, &ncont, conts, ev);
+	if (rc == 0) {
+		*is_truncated = false;
+	} else if (rc == -DER_TRUNC) {
+		rc            = 0;
+		*is_truncated = true;
+	} else {
+		D_ERROR("Failed to list containers in pool, rc = %d\n", rc);
+		rc = daos_der2errno(rc);
+		goto err;
+	}
+
+	for (int i = 0; i < ncont; i++) {
+		char *name = conts[i].pci_label;
+		if (strcmp(name, METADATA_BUCKET) == 0) {
+			D_INFO("Skipping container: %s", name);
+			continue;
+		}
+
+		// Copy bucket name
+		strcpy(buf[bi].name, name);
+
+		// Get info
+		rc = ds3_bucket_open(name, &ds3b, ds3, ev);
+		if (rc != 0) {
+			D_INFO("Skipping container: %s", name);
+			continue;
+		}
+
+		rc = ds3_bucket_get_info(&buf[bi], ds3b, ev);
+		if (rc != 0) {
+			D_INFO("Skipping container: %s", name);
+			rc = ds3_bucket_close(ds3b, ev);
+			continue;
+		}
+
+		rc = ds3_bucket_close(ds3b, ev);
+		if (rc != 0)
+			goto err;
+
+		bi++;
+		// TODO handle marker
+	}
+
+	*nbuck = bi;
+
+err:
+	D_FREE(conts);
+	return -rc;
 }
 
 int
-ds3_bucket_create(const char *name, daos_prop_t *props, ds3_t *ds3, daos_event_t *ev)
+ds3_bucket_create(const char *name, struct ds3_bucket_info *info, dfs_attr_t *attr, ds3_t *ds3,
+		  daos_event_t *ev)
 {
-	return 0;
+	int           rc   = 0;
+	ds3_bucket_t *ds3b = NULL;
+
+	if (ds3 == NULL || name == NULL)
+		return -EINVAL;
+
+	// Prevent attempting to create metadata bucket
+	if (strcmp(name, METADATA_BUCKET) == 0) {
+		D_ERROR("Cannot create metadata bucket");
+		return -EINVAL;
+	}
+
+	// Create dfs container and open ds3b
+	rc = dfs_cont_create_with_label(ds3->poh, name, attr, NULL, NULL, NULL);
+	if (rc != 0) {
+		D_ERROR("Failed to create container, rc = %d\n", rc);
+		return -rc;
+	}
+
+	rc = ds3_bucket_open(name, &ds3b, ds3, ev);
+	if (rc != 0) {
+		D_ERROR("Failed to open container, rc = %d\n", rc);
+		return -rc;
+	}
+
+	rc = ds3_bucket_set_info(info, ds3b, ev);
+	if (rc != 0) {
+		D_ERROR("Failed to put bucket info, rc = %d\n", rc);
+		goto err;
+	}
+
+	// Create multipart index
+	rc = dfs_mkdir(ds3->meta_dfs, ds3->meta_dirs[MULTIPART_DIR], name, DEFFILEMODE, 0);
+	if (rc != 0 && rc != EEXIST) {
+		D_ERROR("Failed to create multipart index, rc = %d\n", rc);
+	}
+
+err:
+	ds3_bucket_close(ds3b, ev);
+	return -rc;
 }
 
 int
-ds3_bucket_destroy(const char *name, ds3_t *ds3, daos_event_t *ev)
+ds3_bucket_destroy(const char *name, bool force, ds3_t *ds3, daos_event_t *ev)
 {
-	return 0;
+	int            rc      = 0;
+	ds3_bucket_t  *ds3b    = NULL;
+	uint32_t       nd      = 10;
+	struct dirent *dirents = NULL;
+	dfs_obj_t     *dir_obj = NULL;
+	daos_anchor_t  anchor;
+
+	if (ds3 == NULL || name == NULL)
+		return -EINVAL;
```

Comment:
We prefer to return negative errno instead of errno (like dfs), right?

Reply:
This is because ceph expects negative errno, but I can change it to positive if folks here prefer positive errno, and I would just add a minus sign when used in ceph